openmls/src/group/tests/external_remove_proposal.rs

```diff
@@ -0,0 +1,327 @@
+    let error = alice_group
+        .process_message(backend, bob_external_remove_proposal)
+        .unwrap_err();
+    assert_eq!(
+        error,
+        ProcessMessageError::ValidationError(ValidationError::UnauthorizedExternalSender)
+    );
+}
+
+#[apply(ciphersuites_and_backends)]
+fn external_remove_proposal_should_fail_with_invalid_external_senders(
+    ciphersuite: Ciphersuite,
+    backend: &impl OpenMlsCryptoProvider,
+) {
+    // delivery service credentials. DS will craft an external remove proposal
+    let ds_credential_bundle = generate_credential_bundle(
+        "delivery-service".into(),
+        ciphersuite.signature_algorithm(),
+        backend,
+    );
+
+    let (mut alice_group, _alice_credential) = validation_test_setup(
+        PURE_PLAINTEXT_WIRE_FORMAT_POLICY,
+        ciphersuite,
+        backend,
+        vec![ExternalSender::new(
+            ds_credential_bundle
+                .credential_with_key
+                .signature_key
+                .clone(),
+            ds_credential_bundle.credential_with_key.credential,
+        )],
+    );
+
+    let ds_invalid_credential_bundle = generate_credential_bundle(
+        "delivery-service-invalid".into(),
+        ciphersuite.signature_algorithm(),
+        backend,
+    );
+
+    // get Bob's index
+    let bob_index = alice_group
+        .members()
+        .find(|member| member.credential.identity() == b"Bob")
+        .map(|member| member.index)
+        .unwrap();
+    // Now Delivery Service wants to (already) remove Bob with invalid sender index
+    let bob_external_remove_proposal: MlsMessageIn = ExternalProposal::new_remove(
+        bob_index,
+        alice_group.group_id().clone(),
+        alice_group.epoch(),
+        &ds_invalid_credential_bundle.signer,
+        0,
+    )
+    .unwrap()
+    .into();
+
+    // Alice tries to validate the message and should fail as sender is invalid
+    let error = alice_group
+        .process_message(backend, bob_external_remove_proposal)
+        .unwrap_err();
+    assert_eq!(error, ProcessMessageError::InvalidSignature);
+}
+
+#[apply(ciphersuites_and_backends)]
+fn external_remove_proposal_should_fail_with_no_external_senders(
+    ciphersuite: Ciphersuite,
+    backend: &impl OpenMlsCryptoProvider,
+) {
+    let (mut alice_group, _) = validation_test_setup(
+        PURE_PLAINTEXT_WIRE_FORMAT_POLICY,
+        ciphersuite,
+        backend,
+        vec![],
+    );
+    // delivery service credentials. DS will craft an external remove proposal
+    let ds_credential_bundle = generate_credential_bundle(
+        "delivery-service".into(),
+        ciphersuite.signature_algorithm(),
+        backend,
+    );
+
+    // get Bob's index
+    let bob_index = alice_group
+        .members()
+        .find(|member| member.credential.identity() == b"Bob")
+        .map(|member| member.index)
+        .unwrap();
+    // Now Delivery Service wants to remove Bob with invalid sender index but there's no extension
```

Comment:
It would be nice to test these two things separately: a) wrong sender index and b) there is no extension.
Ideally also c) existing index but wrong signature.

Reply:
They're already split:
1. external_remove_proposal_should_fail_when_invalid_external_senders_index
2. external_remove_proposal_should_fail_when_no_external_senders
3. external_remove_proposal_should_fail_when_invalid_signature
and external_remove_proposal_should_remove_member for the positive case

Reply:
Sorry, I missed that completely.